Python scripts to automate cleaning up old COPR builds

SilleBille opened the pull request:
COPR nightly builds older than 7 days needs to be cleaned automatically to prevent official COPR from large number of builds

SilleBille committed:
Adding nightly build clean up

Python script to automatically clean up
7 days or older nightly builds is added

SilleBille committed:
Fixed code quality for cleaning up nightly builds

Cleaner way to wipe off old nightly builds in 10.6 COPR

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/pki-copr.py:
```diff
@@ -35,22 +35,19 @@ def print_help():
-	opts, args = getopt.gnu_getopt(sys.argv, 't:r:g:p:', [
-		'token=', 'repo=', 'group=', 'project='])
+	opts, args = getopt.gnu_getopt(sys.argv, 'g:p:', [
+		'group=', 'project='])
```
I think we want to remove -p/--project and keep -r/--repo instead.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+from requests.auth import HTTPBasicAuth
+import datetime
+
+class Utils:
```
Let's rename the class to CoprUtil or something like that.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+            404: "doesn't exists"
+    }
+
+    def __init__(self, repo, copr_url=None):
```
The 'repo' parameter should not be needed here since the repo name will be provided to getProjectID().

edewata replied:
The constructor can take a 'client' parameter which is the COPR API client object. If possible, the copr_url should be obtained from the client parameter.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+        receivedData = self.sendGetRequest(url=url, params=PARAMS)
+
+        # Extract project ID
+        for projectData in receivedData['projects']:
```
Instead of iterating through all returned projects, we need to check to make sure the API returns exactly 1 result. If it returns 0 projects, we should throw 'Project not found' error message. If it returns more than one that means the GET parameters are not specific enough.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+        # Extract project ID
+        for projectData in receivedData['projects']:
+            project = projectData['project']
+            if project['name'] == self.repo:
```
At this point we can assume there's only 1 returned project and the name should match the requested project name, so we don't need to check it again.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+    def getDeleteBuildIDs(self, projectID=None, package=None, days=None):
+        """
+            :param integer projectID: project ID
+            :param unicode package: package name 
```
We don't really need the package parameter since we want to remove all builds older than certain age regardless of the package.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+
+        return buildIDs
+
+    def deleteBuild(self, buildID=None, login=None, token=None):
```
Assuming the 'client' parameter was added in the constructor, it's no longer necessary to provide the login/token parameter here.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,143 @@
+        if name is not None and group is not None:
+            PARAMS = {'group': group, 'name': name}
+        else:
+            return None
```
We should generate an error here because we want to find exactly 1 project, so the caller should provide either owner & name, or group & name.

edewata reviewed (commented):
I have some additional comments.

SilleBille committed:
Improved code quality

- COPR Client now uses V2 API endpoint
- Added generic method to get build IDs based on age
- improved code quality

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,134 @@
+    def __init__(self, client=None):
+        """
+            :param unicode repo: COPR repo to work on
+            :param unicode copr_url: COPR URL to access end points
```
The param docs should match the actual params.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,134 @@
+
+    def getProjectID(self, name=None, group='pki'):
+        """
+            :param unicode searchQuery: search parameter similar to search box in COPR site
```
This param is no longer available.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,134 @@
+    def findBuildIDs(self, projectID=None, package=None, minAge=None):
+        """
+            :param integer projectID: project ID
+            :param unicode package: package name 
```
We want to remove all packages older than certain age, so the package param should not be required.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,134 @@
+        """
+            :param integer buildID: build ID
+            :param unicode login: COPR login parsed from ~/.config/copr
+            :param unicode token: COPR token parsed from ~/.config/copr
```
The param docs should match the actual params.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,134 @@
+
+        return buildIDs
+
+    def deleteBuild(self, buildID=None):
```
buildID is required, so it should not have a default value.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/utils.py:
```diff
@@ -0,0 +1,134 @@
+            :param unicode login: COPR login parsed from ~/.config/copr
+            :param unicode token: COPR token parsed from ~/.config/copr
+        """
+        if buildID is None:
```
This can be simplified into:
> if not buildID:

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/pki-copr.py:
```diff
@@ -0,0 +1,82 @@
+for deleteBuildID in deleteBuildIDs:
+    response = util.deleteBuild(buildID=deleteBuildID)
+    if not response:
+        print("Build ID:", deleteBuildID, " can't be deleted for unknown reasons!")
```
Based on the current code it's impossible to get an empty response, so I think this case should be removed.

edewata commented on jenkins/ansible/roles/createNightlyBuildJob/files/pki-copr.py:
```diff
@@ -0,0 +1,82 @@
+    if not response:
+        print("Build ID:", deleteBuildID, " can't be deleted for unknown reasons!")
+    else:
+        print("Build ID:", deleteBuildID, response)      
```
This should be sufficient for both successful and failed deletion.

edewata reviewed (commented):
I have some additional comments.

SilleBille committed:
Revised documentation
